provider/azure.go

```
@@ -0,0 +1,244 @@
+				},
+			},
+		}, nil
+	}
+	return nil, fmt.Errorf("Invalid record type. Only CNAME, A and TXT are currently supported.")
+}
+
+func (e *azureEndpoint) toRecordSet() (dns.RecordSet, error) {
+	p, err := e.getRecordSetProperties()
+	if err != nil {
+		return dns.RecordSet{}, err
+	}
+	return dns.RecordSet{
+		Name: &e.DNSName,
+		Type: &e.RecordType,
+		RecordSetProperties: p,
+	}, nil
+}
+
+func (r azureRecord) toEndpoints() ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	switch *r.Type {
+	case "CNAME":
+		endpoints = append(endpoints, endpoint.NewEndpoint(*r.Name, *r.RecordSetProperties.CnameRecord.Cname, *r.Type))
+	case "A":
+		for _, ar := range *r.RecordSetProperties.ARecords {
+			endpoints = append(endpoints, endpoint.NewEndpoint(*r.Name, *ar.Ipv4Address, *r.Type))
+		}
+	case "TXT":
+		for _, tr := range *r.RecordSetProperties.TxtRecords {
+			endpoints = append(endpoints, endpoint.NewEndpoint(*r.Name, strings.Join(*tr.Value, " "), *r.Type))
+		}
+	default:
+		return nil, fmt.Errorf("Invalid record type. Only CNAME, A and TXT are currently supported.")
+	}
+
+	return endpoints, nil
+}
+
+func RecordSetListToEndpoints(records []dns.RecordSet) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+	for _, r := range records {
+		a := azureRecord{r}
+		moreEndpoints, err := a.toEndpoints()
+		if err != nil {
+			return nil, err
+		}
+		endpoints = append(endpoints, moreEndpoints...)
+	}
+	return endpoints, nil
+}
+
+
+func (p *azureProvider) List(zone string) ([]dns.RecordSet, error) {
+	var records []dns.RecordSet
+
+	// Get first 100 records
+	res, err := p.recordSetsClient.ListByDNSZone(p.resourceGroupName, zone, nil)
+	if err != nil {
+		return nil, err
+	}
+	records = append(records, *res.Value...)
+	// Get following records by 100
+	for err != nil {
+		res, err = p.recordSetsClient.ListByDNSZone(p.resourceGroupName, zone, nil)
+		if err == nil {
+			records = append(records, *res.Value...)
+		}
+	}
```

I'm not a big fan of silencing errors like this, but maybe it's fine to be best-effort here. Do we expect this to fail often? Otherwise I'd say any error is worth failing the whole call over. That would get rid of the special casing of the first 100 too.